api/src/Service/UcService.php

```diff
@@ -253,8 +253,88 @@ public function validateUserGroups(array $usergroupIds){
+           }
+
+           $employee = [
+               'organization' => $result['@id'],
+               'name' => 'TAALHUIS_EMPLOYEE',
+               'description' => 'UserGroup employee of '.$result['name'],
+           ];
+           if ($update) {
+               $userGroup[] = $this->commonGroundService->updateResource($employee,['component' => 'uc', 'type' => 'groups', 'id' => $userGroupEmployee['id']]);
+           } else {
+               $userGroup[] = $this->commonGroundService->saveResource($employee,['component' => 'uc', 'type' => 'groups']);
+           }
+       } else {
+           if ($update){
+               $userGroupCoordinator = $this->commonGroundService->getResourceList(['component' => 'uc', 'type' => 'groups'], ['organization' => $result['@id']])['hydra:member'][0];
+               $userGroupMentor = $this->commonGroundService->getResourceList(['component' => 'uc', 'type' => 'groups'], ['organization' => $result['@id']])['hydra:member'][1];
+               $userGroupVolunteer = $this->commonGroundService->getResourceList(['component' => 'uc', 'type' => 'groups'], ['organization' => $result['@id']])['hydra:member'][2];
+           }
+           $coordinator = [
+               'organization' => $result['@id'],
+               'name' => 'AANBIEDER_COORDINATOR',
+               'description' => 'UserGroup coordinator of '.$result['name'],
+           ];
+           if ($update) {
+               $userGroup[] = $this->commonGroundService->updateResource($coordinator,['component' => 'uc', 'type' => 'groups', 'id' => $userGroupCoordinator['id']]);
+           } else {
+               $userGroup[] = $this->commonGroundService->saveResource($coordinator,['component' => 'uc', 'type' => 'groups']);
+           }
+
+           $mentor = [
+               'organization' => $result['@id'],
+               'name' => 'AANBIEDER_MENTOR',
+               'description' => 'UserGroup mentor of '.$result['name'],
+           ];
+           if ($update) {
+               $userGroup[] = $this->commonGroundService->updateResource($mentor,['component' => 'uc', 'type' => 'groups', 'id' => $userGroupMentor['id']]);
+           } else {
+               $userGroup[] = $this->commonGroundService->saveResource($mentor,['component' => 'uc', 'type' => 'groups']);
+           }
+
+           $volunteer = [
+               'organization' => $result['@id'],
+               'name' => 'AANBIEDER_VOLUNTEER',
+               'description' => 'UserGroup volunteer of '.$result['name'],
+           ];
+           if ($update) {
+               $userGroup[] = $this->commonGroundService->updateResource($volunteer,['component' => 'uc', 'type' => 'groups', 'id' => $userGroupVolunteer['id']]);
+           } else {
+               $userGroup[] = $this->commonGroundService->saveResource($volunteer,['component' => 'uc', 'type' => 'groups']);
+           }
+       }
+       return $userGroup;
+    }
```

Comment:
Okay, this actually can be split up a lot better. Maybe a good rule to follow, if possible, is to split your code in a new function once you have to nest multiple logic operators (if, foreach, while, switch, etc).

As an example for further function splitting I have reworked this code in the suggestion below. This is not yet perfect (for example, the Coordinator-functions could be combined again using the $type parameter, but that would introduce some code complexity that is not beneficial to the example I'm trying to make.

```suggestion
    public function createTaalhuisCoordinatorGroup(array $result, array $userGroups, ?array $userGroupCoordinator = null): array
    {
        $coordinator = [
            'organization' => $result['@id'],
            'name' => 'TAALHUIS_COORDINATOR',
            'description' => 'UserGroup coordinator of '.$result['name'],
        ];
        if ($userGroupCoordinator) {
            $userGroups[] = $this->commonGroundService->updateResource($coordinator,['component' => 'uc', 'type' => 'groups', 'id' => $userGroupCoordinator['id']]);
        } else {
            $userGroups[] = $this->commonGroundService->saveResource($coordinator,['component' => 'uc', 'type' => 'groups']);
        }
        return $userGroups;
    }

    public function createTaalhuisEmployeeGroup (array $result, array $userGroups, ?array $userGroupEmployee = null): array
    {
        $employee = [
            'organization' => $result['@id'],
            'name' => 'TAALHUIS_EMPLOYEE',
            'description' => 'UserGroup employee of '.$result['name'],
        ];
        if ($userGroupEmployee) {
            $userGroups[] = $this->commonGroundService->updateResource($employee,['component' => 'uc', 'type' => 'groups', 'id' => $userGroupEmployee['id']]);
        } else {
            $userGroups[] = $this->commonGroundService->saveResource($employee,['component' => 'uc', 'type' => 'groups']);
        }
        return $userGroups;
    }

    public function createTaalhuizenUserGroups(array $result, array $userGroups, bool $update): array
    {
        if ($update){
            $userGroupCoordinator = $this->commonGroundService->getResourceList(['component' => 'uc', 'type' => 'groups'], ['organization' => $result['@id']])['hydra:member'][0];
            $userGroupEmployee = $this->commonGroundService->getResourceList(['component' => 'uc', 'type' => 'groups'], ['organization' => $result['@id']])['hydra:member'][1];
        } else {
            $userGroupCoordinator = null;
            $userGroupEmployee = null;
        }
        $userGroups = $this->createTaalhuisCoordinatorGroup($result, $userGroups, $userGroupCoordinator);
        $userGroups = $this->createTaalhuisEmployeeGroup($result, $userGroups, $userGroupEmployee);
        return $userGroups;
    }
    
    public function createProviderCoordinatorUserGroup(array $result, array $userGroups, ?array $userGroupCoordinator = null): array
    {
        $coordinator = [
            'organization' => $result['@id'],
            'name' => 'AANBIEDER_COORDINATOR',
            'description' => 'UserGroup coordinator of '.$result['name'],
        ];
        if ($userGroupCoordinator) {
            $userGroups[] = $this->commonGroundService->updateResource($coordinator,['component' => 'uc', 'type' => 'groups', 'id' => $userGroupCoordinator['id']]);
        } else {
            $userGroups[] = $this->commonGroundService->saveResource($coordinator,['component' => 'uc', 'type' => 'groups']);
        }
        return $userGroups;
    }
    
    public function createProviderMentorUserGroup(array $result, array $userGroups, ?array $userGroupMentor = null): array
    {
        $mentor = [
            'organization' => $result['@id'],
            'name' => 'AANBIEDER_MENTOR',
            'description' => 'UserGroup mentor of '.$result['name'],
        ];
        if ($userGroupMentor) {
            $userGroups[] = $this->commonGroundService->updateResource($mentor,['component' => 'uc', 'type' => 'groups', 'id' => $userGroupMentor['id']]);
        } else {
            $userGroups[] = $this->commonGroundService->saveResource($mentor,['component' => 'uc', 'type' => 'groups']);
        }
        return $userGroups;
    }
    
    public function createProviderVolunteerUserGroup(array $result, array $userGroups, ?array $userGroupVolunteer = null): array
    {
        $volunteer = [
            'organization' => $result['@id'],
            'name' => 'AANBIEDER_VOLUNTEER',
            'description' => 'UserGroup volunteer of '.$result['name'],
        ];
        if ($userGroupVolunteer) {
            $userGroups[] = $this->commonGroundService->updateResource($volunteer,['component' => 'uc', 'type' => 'groups', 'id' => $userGroupVolunteer['id']]);
        } else {
            $userGroups[] = $this->commonGroundService->saveResource($volunteer,['component' => 'uc', 'type' => 'groups']);
        }
        return $userGroups;
    }

    public function createProviderUserGroups(array $result, array $userGroups, bool $update): array
    {
        if ($update){
            $userGroupCoordinator = $this->commonGroundService->getResourceList(['component' => 'uc', 'type' => 'groups'], ['organization' => $result['@id']])['hydra:member'][0];
            $userGroupMentor = $this->commonGroundService->getResourceList(['component' => 'uc', 'type' => 'groups'], ['organization' => $result['@id']])['hydra:member'][1];
            $userGroupVolunteer = $this->commonGroundService->getResourceList(['component' => 'uc', 'type' => 'groups'], ['organization' => $result['@id']])['hydra:member'][2];
        } else {
            $userGroupCoordinator = null;
            $userGroupMentor = null;
            $userGroupVolunteer = null;
        }
        $this->createProviderCoordinatorUserGroup($result, $userGroups, $userGroupCoordinator);
        $this->createProviderMentorUserGroup($result, $userGroups, $userGroupMentor);
        $this->createProviderVolunteerUserGroup($result, $userGroups, $userGroupVolunteer);

        return $userGroups;
    }

    public function createUserGroups(array $result, $type, $update = false): array
    {
        $userGroups = [];
        if ($type == 'Taalhuis') {
           $userGroups = $this->createTaalhuizenUserGroups($result, $userGroups, $update);
        } else {
           $userGroups = $this->createProviderUserGroups($result, $userGroups, $update)
        }
        return $userGroups;
    }
```

You see that this rather complicated function falls apart in 8 extremely simple functions, that can be reused in other functions if needed. This benefits the maintainability of your code, and also the readability, otherwise it would rede like "if this and if that, but not that", while this code reads like "if this then that", and then you jump to the next piece of code that again reads "if this then that".